### PR TITLE
Code fix add missing method

### LINKED ISSUE #14329: Add Missing Property QuickFix on Method Call Should Create Method
**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

```ts
class Foo {
    constructor() {
        this.bar('abc', 123)
    }
}
```

Get a quick fixes for the call to `bar`. Trigger the quick fix to add a missing property to fix the error

screen shot 2017-02-26 at 11 26 52 pm

**Expected behavior:**
Fix introduces a method that fixes the error

```ts
class Foo {
     bar(p1: string, p2: number) {
          throw new Error('Method not implemented.');
     }

    constructor() {
        this.bar('abc', 123)
    }
}
```

**Actual behavior:**
Fix introduces a `any` property that fixes the error but is not very helpful:

```ts
class Foo {
    bar: any
    constructor() {
        this.bar('abc', 123)
    }
}
```

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes #14329. Changes to fourslash based on https://github.com/Microsoft/TypeScript/pull/14530.

### CODE REVIEW

File: src/services/codefixes/fixAddMissingMember.ts [hunk 1]
```diff
@@ -66,47 +65,57 @@ namespace ts.codefix {
...
-            let typeNode: TypeNode;
+            const openBrace = getOpenBraceOfClassLike(classDeclaration, sourceFile);
+            let actions: CodeAction[];
+
+            if (token.parent.parent.kind === SyntaxKind.CallExpression) {
+                const callExpression = <CallExpression>token.parent.parent;
+                const methodDeclaration = createMethodFromCallExpression(callExpression, tokenName);
+
+                const methodDeclarationChangeTracker = textChanges.ChangeTracker.fromCodeFixContext(context);
+                methodDeclarationChangeTracker.insertNodeAfter(sourceFile, openBrace, methodDeclaration, { suffix: context.newLineCharacter });
```
Comment (mhegazy) [inline src/services/codefixes/fixAddMissingMember.ts]: add handling for `static` methods.

Comment (mhegazy) [inline src/services/codefixes/fixAddMissingMember.ts]: also add handling for JS functions. basically the same but no type annotation or type parameters.

Comment (mhegazy) [inline src/services/codefixes/fixAddMissingMember.ts]: also please add tests for these.

File: tests/cases/fourslash/codeFixUndeclaredMethod.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+verify.applyCodeFix(/*errorCode*/undefined, 0);
+verify.applyCodeFix(/*errorCode*/undefined, 0);
+verify.applyCodeFix(/*errorCode*/undefined, 0);
+
+verify.rangeIs(`
+    foo3<T0, T1, T2, T3, T4, T5, T6, T7>() {
+        throw new Error("Method not implemented.");
+    }
+    foo2<T, U, V, W, X, Y, Z>() {
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/codeFixUndeclaredMethod.ts]: Can you add a test for when you need a letter past `Z`? 😄

Comment (aozgaa) [inline tests/cases/fourslash/codeFixUndeclaredMethod.ts]: That is what `foo3` is, in the existing test.

File: src/services/codefixes/helpers.ts [hunk 3]
```diff
@@ -142,6 +142,46 @@ namespace ts.codefix {
...
+                /*modifiers*/ undefined,
+                /*dotDotDotToken*/ undefined,
+                `arg${i}`,
+                /*questionToken*/ undefined,
+                typeNode,
+                /*initializer*/ undefined);
+            parameters.push(newParameter);
+        }
+
+        const typeArgCount = callExpression.typeArguments ? callExpression.typeArguments.length : 0;
```
Comment (rbuckton) [inline src/services/codefixes/helpers.ts]: `length(callExpression.typeArguments)`

File: src/services/codefixes/helpers.ts [hunk 4]
```diff
@@ -142,6 +142,46 @@ namespace ts.codefix {
         }
     }
 
+    export function createMethodFromCallExpression(callExpression: CallExpression, methodName: string): MethodDeclaration {
+        const argCount = callExpression.arguments.length;
+        const parameters: ParameterDeclaration[] = [];
+        for (let i = 0; i < argCount; i++) {
+            const typeNode = createKeywordTypeNode(SyntaxKind.AnyKeyword);
```
Comment (rbuckton) [inline src/services/codefixes/helpers.ts]: Do not include types in JavaScript files.

File: src/services/codefixes/fixAddMissingMember.ts [hunk 5]
```diff
@@ -134,15 +178,32 @@ namespace ts.codefix {
...
                 });
             }
 
             return actions;
         }
+
+        function getActionForMethodDeclaration(): CodeAction | undefined {
+            if (token.parent.parent.kind === SyntaxKind.CallExpression) {
+                const callExpression = <CallExpression>token.parent.parent;
+                const methodDeclaration = createMethodFromCallExpression(callExpression, tokenName, /*includeTypeScriptSyntax*/ true, makeStatic);
```
Comment (rbuckton) [inline src/services/codefixes/fixAddMissingMember.ts]: You call this from both `getActionsForAddMissingMemberInJavaScriptFile` and `getActionsForAddMissingMemberInTypeScriptFile`. Do you need to pass in a different value for `includeTypeScriptSyntax` here based on the caller?

File: src/services/codefixes/fixAddMissingMember.ts [hunk 6]
```diff
@@ -8,107 +8,160 @@ namespace ts.codefix {
...
+
+            const checker = context.program.getTypeChecker();
+            const leftExpression = token.parent.expression;
+            const leftExpressionType = checker.getTypeAtLocation(leftExpression);
+
+            if (leftExpressionType.flags & TypeFlags.Object) {
+                const symbol = leftExpressionType.symbol;
+                if (symbol.flags & SymbolFlags.Class) {
+                    classDeclaration = symbol.declarations && <ClassLikeDeclaration>symbol.declarations[0];
+                    if (getObjectFlags(leftExpressionType) & ObjectFlags.Anonymous && symbol.flags & SymbolFlags.Class && !checker.getBaseTypeVariableOfClass(symbol)) {
```
Comment (mhegazy) [inline src/services/codefixes/fixAddMissingMember.ts]: not sure i understand this line.

Comment (aozgaa) [inline src/services/codefixes/fixAddMissingMember.ts]: Under these conditions, the LHS of the qualified access expression is the static side of some class. The second check is redundant, and I'll remove it.

Is there a better way to check that the LHS corresponds to the static side of a class?

Comment (aozgaa) [inline src/services/codefixes/fixAddMissingMember.ts]: The logic for the check comes from `createAnonymousTypeNode` in checker.ts.

File: src/compiler/utilities.ts [hunk 7]
```diff
@@ -733,6 +733,10 @@ namespace ts {
         return false;
     }
 
+    export function getObjectFlags(type: Type): ObjectFlags {
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: can we get that back in

**Pull request merged**